core/nightshade/src/nightshade.rs

```diff
@@ -0,0 +1,781 @@
+        self.committed.is_some()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rand::{SeedableRng, XorShiftRng};
+
+    use primitives::aggregate_signature::AggregateSignature;
+
+    use super::*;
+
+    fn generate_bls_key_pairs(total: usize) -> (Vec<BlsPublicKey>, Vec<BlsSecretKey>) {
+        // Use rng to create deterministic tests
+        let mut rng = XorShiftRng::from_seed([11111111, 22222222, 33333333, 44444444]);
+
+        (0..total).map(|_| {
+            let secret_key = BlsSecretKey::generate_from_rng(&mut rng);
+            let public_key = secret_key.get_public_key();
+            (public_key, secret_key)
+        }).unzip()
+    }
+
+    fn check_state_proofs(state: &State) {
+        assert_eq!(state.bare_state.primary_confidence == 0, state.primary_proof.is_none());
+        assert_eq!(state.bare_state.secondary_confidence == 0, state.secondary_proof.is_none());
+    }
+
+    fn header(author: AuthorityId) -> BlockHeader {
+        BlockHeader {
+            author,
+            hash: hash_struct(&author),
+        }
+    }
+
+    #[test]
+    fn bls_on_bare_states() {
+        let (pks, sks) = generate_bls_key_pairs(2);
+        let triplet = BareState {
+            primary_confidence: 0,
+            endorses: header(1),
+            secondary_confidence: 0,
+        };
+        // Aggregate signature
+        let mut aggregated_signature = AggregateSignature::new();
+        for sk in sks {
+            let s = triplet.sign(&sk);
+            aggregated_signature.aggregate(&s);
+        }
+        let signature = aggregated_signature.get_signature();
+        // Aggregate public keys
+        let mut aggregated_pk = AggregatePublicKey::new();
+        for pk in pks {
+            aggregated_pk.aggregate(&pk);
+        }
+        let a_pk = aggregated_pk.get_key();
+
+        assert_eq!(a_pk.verify(&triplet.bs_encode(), &signature), true);
+    }
+
+    fn create_nightshades(num_authorities: usize) -> Vec<Nightshade> {
+        let (public_keys, secret_keys) = generate_bls_key_pairs(num_authorities);
+
+        let ns: Vec<_> = (0..num_authorities)
+            .map(|i|
+                Nightshade::new(
+                    i,
+                    num_authorities,
+                    header(i),
+                    public_keys.clone(),
+                    secret_keys[i].clone(),
+                )
+            ).collect();
+
+        ns
+    }
+
+    fn nightshade_all_sync(num_authorities: usize, num_rounds: usize) {
+        let mut ns = create_nightshades(num_authorities);
+
+        for _ in 0..num_rounds {
+            let mut states = vec![];
+
+            for i in 0..num_authorities {
+                let state = ns[i].state();
+                check_state_proofs(&state);
+                states.push(state);
+            }
+
+            for i in 0..num_authorities {
+                for j in 0..num_authorities {
+                    if i != j {
+                        ns[i].update_state(j, states[j].clone());
```

Comment:
Why result of `update_state` is not used?

Reply:
fixed